### PR TITLE
Ledger assignment based on outstanding value

### PR DESCRIPTION
**Pull request by johnwatson484**
https://eaflood.atlassian.net/browse/SFI-1625﻿

### CODE REVIEW

File: app/processing/delta/get-outstanding-ledger-values.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const { AP, AR } = require('../../ledgers')
+
+const getOutstandingLedgerValues = (previousPaymentRequests) => {
```
Comment (johnwatson484) [inline app/processing/delta/get-outstanding-ledger-values.js:3]: I've opted to rename references to `unsettled` as it's no longer appropriate given we now know that we don't need to worry about the settlement status of AR.  We just need to know what debt is outstanding.  Unsettled values is specific to AP.

So I've chosen `getOutstandingLedgerValues` as essentially that's what we need to find out.  ie, how much is outstanding on each ledger that we should be looking to cancel out.

File: app/settlement.js [hunk 2]
```diff
@@ -2,8 +2,9 @@ const db = require('./data')
 
 const updateSettlementStatus = async (returnData) => {
   if (returnData.settled) {
-    await db.completedPaymentRequest.update({ settled: returnData.settlementDate }, { where: { invoiceNumber: returnData.invoiceNumber } })
+    await db.completedPaymentRequest.update({ lastSettlement: returnData.settlementDate, settledValue: db.Sequelize.literal(`COALESCE("settledValue", 0) + ${returnData.value}`) }, { where: { invoiceNumber: returnData.invoiceNumber } })
```
Comment (johnwatson484) [inline app/settlement.js:5]: I've used COALESCE to ensure that there is a value of 0 if `settledValue` is null.  That shouldn't be the case but just as a failsafe to ensure that the return value can't be ignored.

File: app/data/models/completed-payment-request.js [hunk 3]
```diff
@@ -24,7 +24,8 @@ module.exports = (sequelize, DataTypes) => {
     value: DataTypes.INTEGER,
     submitted: DataTypes.DATE,
     acknowledged: DataTypes.DATE,
-    settled: DataTypes.DATE,
+    lastSettlement: DataTypes.DATE,
+    settledValue: DataTypes.INTEGER,
```
Comment (marctemp) [inline app/data/models/completed-payment-request.js:28]: Is this guaranteed to be an int rather that float?

File: app/settlement.js (hunk 2 above)
Comment (marctemp) [inline app/settlement.js:5]: Just confirming, `settled` is still returned from DAX we are just adding the extra field `value` to that return file

File: test/unit/processing/delta/reallocate-to-ledger.test.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ describe('allocate to ledgers', () => {
     expect(updatedPaymentRequests[0].ledger).toBe(AR)
   })
 
-  test('should reallocate all AR to AP if unsettled value is greater than current request recovery', () => {
```
Comment (marctemp) [inline test/unit/processing/delta/reallocate-to-ledger.test.js:14]: Should this description match the one above but to `all AR to AP`? It's describing a valid scenario in which this switch will happen but the functionality doesn't look/compute so is a bit of a red heron

File: test/unit/processing/delta/calculate-delta.test.js [hunk 5]
```diff
@@ -259,7 +260,7 @@ describe('calculate delta', () => {
     const previousPaymentRequests = [{
       ledger: AP,
       value: 200,
-      settled: new Date(2022, 3, 5),
```
Comment (marctemp) [inline test/unit/processing/delta/calculate-delta.test.js:262]: Should this not be `settledValue: 200` to correspond with test above?

File: test/unit/processing/delta/calculate-delta.test.js [hunk 6]
```diff
@@ -259,7 +260,7 @@ describe('calculate delta', () => {
...
+      settledValue: 0,
```
Comment (marctemp) [inline test/unit/processing/delta/calculate-delta.test.js:263]: Should this not be settledValue: 200 to correspond with test above?

**Review by marctemp — CHANGES_REQUESTED** [review]
Just minor questions/changes

File: app/data/models/completed-payment-request.js (hunk 3 above)
Comment (johnwatson484) [reply app/data/models/completed-payment-request.js:28]: Yes.  The `ffc-pay-responses` repo will take the decimal value from the actual return file and convert it to pence.

File: test/unit/processing/delta/calculate-delta.test.js [hunk 7]
```diff
@@ -42,7 +42,7 @@ describe('calculate delta', () => {
     const previousPaymentRequests = [{
       ledger: AP,
       value: 80,
-      settled: new Date(2022, 3, 5),
+      settledValue: 80,
```
[the remark this answers is not in the document]
Comment (johnwatson484) [reply test/unit/processing/delta/calculate-delta.test.js:45]: Good spot.  Fixed

File: test/unit/processing/delta/calculate-delta.test.js (hunk 6 above)
Comment (johnwatson484) [reply test/unit/processing/delta/calculate-delta.test.js:263]: It technically doesn't change the scope of the test as we're not testing ledger assignment but I agree it reduces the variables if we assume fully settled and we're consistent with the equivalent test as you say.  Updated.

File: test/unit/processing/delta/reallocate-to-ledger.test.js (hunk 4 above)
Comment (johnwatson484) [reply test/unit/processing/delta/reallocate-to-ledger.test.js:14]: Fair point.  I actually think the above test is named incorrectly too.

The intention of the test is to see if the ledger changes.  The reason why the ledger changes is outside the scope so no need to mention it and the reasons are actually covered in higher level tests already.  I've therefore renamed both these tests.

**Review by marctemp — APPROVED** [review]
All improvements suggested and pushed, LGTM

**Pull request merged**